Mathlib/Order/SuccPred/InverseSystem.lean

```diff
@@ -0,0 +1,294 @@
+    have hi : i ∈ s := hs (le_succ i) hi
+    rw [piLTProj_intro (f := e₁ _ x) (le_succ i) (by exact hj),
+        ← nat₁ _ hi (by exact le_succ i), ih, nat₂ _ hi (by exact le_succ i)]
+  · ext x j
+    have ⟨k, hjk, hki⟩ := lim.mid j.2
+    have hk : k ∈ s := hs hki.le hi
+    rw [piLTProj_intro (f := e₁ _ x) hki.le hjk, piLTProj_intro (f := e₂ _ x) hki.le hjk,
+      ← nat₁ _ hk, ← nat₂ _ hk, ih _ hki]
+
+theorem pEquivOn_apply_eq (h : IsLowerSet (s ∩ t))
+    {e₁ : PEquivOn f equivSucc s} {e₂ : PEquivOn f equivSucc t} {i} {his : i ∈ s} {hit : i ∈ t} :
+    e₁.equiv ⟨i, his⟩ = e₂.equiv ⟨i, hit⟩ :=
+  show (e₁.restrict <| inter_subset_left).equiv ⟨i, his, hit⟩ =
+       (e₂.restrict <| inter_subset_right).equiv ⟨i, his, hit⟩ from
+  congr_fun (congr_arg _ <| unique_pEquivOn h) _
+
+/-- Extend a partial family of bijections by one step. -/
+def pEquivOnSucc [InverseSystem f] (hi : ¬IsMax i) (e : PEquivOn f equivSucc (Iic i))
+    (H : ∀ ⦃i⦄ (hi : ¬ IsMax i) x, (equivSucc hi x).1 = f (le_succ i) x) :
+    PEquivOn f equivSucc (Iic (i⁺)) where
+  equiv := piEquivSucc e.equiv (equivSucc hi) hi
+  nat := isNatEquiv_piEquivSucc hi (H hi) e.nat
+  compat hsj hj x := by
+    obtain eq | lt := hsj.eq_or_lt
+    · cases (succ_eq_succ_iff_of_not_isMax hj hi).mp eq; simp [piEquivSucc]
+    · rwa [piEquivSucc, piSplitLE_lt, e.compat]
+
+variable (hi : IsSuccPrelimit i) (e : ∀ j : Iio i, PEquivOn f equivSucc (Iic j))
+
+/-- Glue partial families of bijections at a limit ordinal,
+obtaining a partial family over an open interval. -/
+noncomputable def pEquivOnGlue : PEquivOn f equivSucc (Iio i) where
+  equiv := (piLTLim (X := fun j ↦ F j ≃ piLT X j) hi).symm
+    ⟨fun j ↦ ((e j).restrict fun _ h ↦ h.le).equiv, fun _ _ h ↦ funext fun _ ↦
+      pEquivOn_apply_eq ((isLowerSet_Iio _).inter <| isLowerSet_Iio _)⟩
+  nat j k hj hk h := by rw [piLTLim_symm_apply]; exacts [(e _).nat _ _ _, h.trans_lt (hi.mid _).2.1]
+  compat hj := have k := hi.mid hj
+    by rw [piLTLim_symm_apply hi ⟨_, k.2.2⟩ (by exact k.2.1)]; apply (e _).compat
+
+/-- Extend `pEquivOnGlue` by one step, obtaining a partial family over a closed interval. -/
+noncomputable def pEquivOnLim [InverseSystem f]
+    (equivLim : F i ≃ limit f i) (H : ∀ x l, (equivLim x).1 l = f l.2.le x) :
+    PEquivOn f equivSucc (Iic i) where
+  equiv := piEquivLim (pEquivOnGlue hi e).nat equivLim hi
+  nat := isNatEquiv_piEquivLim (pEquivOnGlue hi e).nat hi H
+  compat hsj hj x := by
+    rw [piEquivLim, piSplitLE_lt (hi.succ_lt <| (succ_le_iff_of_not_isMax hj).mp hsj)]
+    apply (pEquivOnGlue hi e).compat
+
+end Unique
+
+/-- Over a well-ordered type, construct a family of bijections by transfinite recursion. -/
+noncomputable def globalEquiv [WellFoundedLT ι] [SuccOrder ι] [InverseSystem f]
+    (equivSucc : ∀ i, ¬IsMax i → {e : F (i⁺) ≃ F i × X i // ∀ x, (e x).1 = f (le_succ i) x})
+    (equivLim : ∀ i, IsSuccPrelimit i → {e : F i ≃ limit f i // ∀ x l, (e x).1 l = f l.2.le x})
+    (i : ι) : F i ≃ piLT X i :=
```

Comment:
IIUC this commutes with the projection maps, right? Can you prove it?

Reply:
Yes, I was just lazy ...

Reply:
Do you want to do it in this PR? Otherwise I'm happy with leaving it as a TODO

Reply:
Done in [b6bdb35](https://github.com/leanprover-community/mathlib4/pull/18067/commits/b6bdb35215743314c067d33570fd00aa6a15e440); just requires a bit of refactoring. A bit unexpected though that I need to use `pEquivOn_apply_eq` (i.e. uniqueness of `pEquivOn`) once more.